polkadot/node/core/prospective-parachains/src/lib.rs

```diff
@@ -430,81 +455,121 @@ async fn handle_candidate_introduced<Context>(
 				"Received seconded candidate for inactive para",
 			);
 
-			let _ = tx.send(Vec::new());
-			return Ok(())
+			let _ = tx.send(false);
+			return
 		},
 		Some(storage) => storage,
 	};
 
-	let candidate_hash = match storage.add_candidate(candidate, pvd) {
-		Ok(c) => c,
-		Err(CandidateStorageInsertionError::CandidateAlreadyKnown(c)) => {
-			// Candidate known - return existing fragment tree membership.
-			let _ = tx.send(fragment_tree_membership(&view.active_leaves, para, c));
-			return Ok(())
-		},
-		Err(CandidateStorageInsertionError::PersistedValidationDataMismatch) => {
-			// We can't log the candidate hash without either doing more ~expensive
-			// hashing but this branch indicates something is seriously wrong elsewhere
-			// so it's doubtful that it would affect debugging.
+	let actual_relay_parent = candidate.descriptor.relay_parent;
+	let parent_head_hash = pvd.parent_head.hash();
+	let output_head_hash = Some(candidate.commitments.head_data.hash());
+	let candidate_hash =
+		match storage.add_candidate(candidate.clone(), pvd, CandidateState::Seconded) {
+			Ok(c) => c,
+			Err(CandidateStorageInsertionError::CandidateAlreadyKnown(_)) => {
+				gum::warn!(
+					target: LOG_TARGET,
+					para = ?para,
+					"Attempting to introduce an already known candidate: {:?}",
+					candidate.hash()
+				);
+				// Candidate already known.
+				let _ = tx.send(true);
+				return
+			},
+			Err(CandidateStorageInsertionError::PersistedValidationDataMismatch) => {
+				// We can't log the candidate hash without either doing more ~expensive
+				// hashing but this branch indicates something is seriously wrong elsewhere
+				// so it's doubtful that it would affect debugging.
 
-			gum::warn!(
+				gum::warn!(
+					target: LOG_TARGET,
+					para = ?para,
+					"Received seconded candidate had mismatching validation data",
+				);
+
+				let _ = tx.send(false);
+				return
+			},
+		};
+
+	let mut keep_in_storage = false;
+	for (relay_parent, leaf_data) in view.active_leaves.iter_mut() {
+		if let Some(chain) = leaf_data.fragment_chains.get_mut(&para) {
+			gum::debug!(
```

Comment:
```suggestion
			gum::trace!(
```

Reply:
these are still only draft logs. I have a bulletpoint in the PR description to add/fix logs